model/CompatibilityChecker.php

```diff
@@ -14,101 +14,222 @@
  * along with this program; if not, write to the Free Software
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  *
- * Copyright (c) 2015-2019 (original work) Open Assessment Technologies SA;
- *
+ * Copyright (c) 2015-2021 (original work) Open Assessment Technologies SA;
  */
 
 namespace oat\taoClientDiagnostic\model;
 
+use oat\oatbox\service\ConfigurableService;
+use oat\taoClientDiagnostic\model\diagnostic\DiagnosticServiceInterface;
+use oat\taoClientDiagnostic\model\exclusionList\ExcludedBrowserClassService;
+use oat\taoClientDiagnostic\model\exclusionList\ExcludedOsClassService;
 use Sinergi\BrowserDetector\Browser;
 use Sinergi\BrowserDetector\Os;
+use tao_models_classes_FileNotFoundException;
 
-class CompatibilityChecker
+class CompatibilityChecker extends ConfigurableService
 {
+    public const SERVICE_ID = 'taoClientDiagnostic/CompatibilityChecker';
+
+    public const COMPATIBILITY_NONE = 0;
+    public const COMPATIBILITY_COMPATIBLE = 1;
+    public const COMPATIBILITY_NOT_TESTED = 2;
+    public const COMPATIBILITY_SUPPORTED = 3;
+    public const COMPATIBILITY_NOT_SUPPORTED = 4;
+
     protected $compatibility;
+    protected $supported;
+    protected $excludedBrowsers;
+    protected $excludedOS;
 
     /**
-     * CompatibilityChecker constructor
-     * Check parameter required
      * Extract compatibility file
-     * @throws \tao_models_classes_FileNotFoundException
+     * @throws tao_models_classes_FileNotFoundException
+     */
+    public function getCompatibilityList(): array
+    {
+        if (!$this->compatibility) {
+            $compatibilityFile = __DIR__ . '/../include/compatibility.json';
+
+            if (!file_exists($compatibilityFile)) {
+                throw new tao_models_classes_FileNotFoundException("Unable to find the compatibility file");
+            }
+            $this->compatibility = json_decode(file_get_contents($compatibilityFile), true);
+        }
+        return $this->compatibility;
+    }
+
+    /**
+     * Fetch the support list
+     * @throws tao_models_classes_FileNotFoundException
+     */
+    public function getSupportedList(): array
+    {
+        if (!$this->supported) {
+            $service = $this->getServiceLocator()->get(DiagnosticServiceInterface::SERVICE_ID);
+            $config = $service->getDiagnosticJsConfig();
+            $supportListUrl = $config['diagnostic']['testers']['browser']['browserslistUrl'];
+
+            if (!$supportListUrl) {
+                throw new tao_models_classes_FileNotFoundException("The URL to the list of supported browser is not configured");
+            }
+            $supportedList = json_decode(file_get_contents($supportListUrl), true);
+
+            if (!$supportedList) {
```

Comment:
Let's flush a log record on failure with the reason what happened, it is Github down or json_decode error etc

Reply:
Is not the exception caught and logged?
We don't have so many info from the `json_decode` nor the `file_get_contents` unfortunately, to better identify the problem. But I've split the code in order to separate read and decode issues.

Reply:
in a situation with http error codes in `file_get_contents`, it won't reach our `common_exception_InconsistentData`  having been broken on `file_get_contents()` with a warning, example:

`Warning: file_get_contents(http://gifthub.co): Failed to open stream: HTTP request failed! HTTP/1.1 403 Forbidden
`

`Warning: file_get_contents(http://gifthub.com4): Failed to open stream: php_network_getaddresses: getaddrinfo failed: nodename nor servname provided, or not known in Command line code on line 1`

So the idea is to put file_get_contents in a try-cache and then throw a TAO exception that will get to the log.
It's a bit tricky because those warnings can't be handled unless the function is called in 'suppression mode' (`@` operator).

Let's try this instead of try-cache:

```
$supportedListData = @file_get_contents($supportListUrl);
if ($supportedListData === null) {
    $error = error_get_last();
    throw new tao_models_classes_FileNotFoundException(sprintf(
        'Unable to fetch the list of supported browsers due to error: %s',
        $error['message']
    ));
}

$supportedList = json_decode($supportedListData, true);

if (!$supportedList) {
    throw new common_exception_InconsistentData('Unable to decode list of supported browsers');
}
```